src/snowflake/snowpark/_internal/udf_utils.py

```diff
@@ -977,6 +978,322 @@ def resolve_imports_and_packages(
     )
 
 
+def resolve_packages(
+    session: "snowflake.snowpark.Session",
+    packages: Optional[List[Union[str, ModuleType]]],
+    is_pandas_udf: bool = False,
+    statement_params: Optional[Dict[str, str]] = None,
+) -> str:
+
+    # resolve packages
+    resolved_packages = (
+        session._resolve_packages(
+            packages, include_pandas=is_pandas_udf, statement_params=statement_params
+        )
+        if packages is not None
+        else session._resolve_packages(
+            [],
+            session._packages,
+            validate_package=False,
+            include_pandas=is_pandas_udf,
+            statement_params=statement_params,
+        )
+    )
+
+    all_packages = ",".join([f"'{package}'" for package in resolved_packages])
+    return all_packages
+
+
+def resolve_imports(
+    session: "snowflake.snowpark.Session",
+    stage_location: Optional[str],
+    callableProperties: CallableProperties,
+    imports: Optional[List[Union[str, Tuple[str, str]]]],
+    arg_names: List[str],
+    statement_params: Optional[Dict[str, str]] = None,
+    is_dataframe_input: bool = False,
+    skip_upload_on_content_match: bool = False,
+    is_pandas_udf: bool = False,
+) -> Tuple[str, str, str, str, bool]:
+
```

Comment:
`session.add_imports` performs a lazy execution, i.e. only adds imports provided by the user to a running list, and does not actually interact with Snowflake. With the CLI local sandbox, we will do something similar here, instead of the implementation below.